client/Windows/wf_cliprdr.c

```diff
@@ -1975,41 +1975,70 @@ void wf_cliprdr_init(wfContext* wfc, CliprdrClientContext* cliprdr)
+	if (!(clipboard->format_mappings = (formatMapping*) calloc(1, sizeof(formatMapping) * clipboard->map_capacity)))
+		goto fail_format_mappings;
 
 	clipboard->file_array_size = 32;
-	clipboard->file_names = (WCHAR**) calloc(1, clipboard->file_array_size * sizeof(WCHAR*));
-	clipboard->fileDescriptor = (FILEDESCRIPTORW**) calloc(1, clipboard->file_array_size * sizeof(FILEDESCRIPTORW*));
+	if (!(clipboard->file_names = (WCHAR**) calloc(1, clipboard->file_array_size * sizeof(WCHAR*))))
+		goto fail_file_names;
+
+	if (!(clipboard->fileDescriptor = (FILEDESCRIPTORW**) calloc(1, clipboard->file_array_size * sizeof(FILEDESCRIPTORW*))))
+		goto fail_file_descriptor;
 
-	clipboard->response_data_event = CreateEvent(NULL, TRUE, FALSE, _T("response_data_event"));
+	if (!(clipboard->response_data_event = CreateEvent(NULL, TRUE, FALSE, _T("response_data_event"))))
+		goto fail_data_event;
+
+	if (!(clipboard->req_fevent = CreateEvent(NULL, TRUE, FALSE, _T("request_filecontents_event"))))
+		goto fail_filecontents_event;
 
-	clipboard->req_fevent = CreateEvent(NULL, TRUE, FALSE, _T("request_filecontents_event"));
 	clipboard->ID_FILEDESCRIPTORW = RegisterClipboardFormatW(CFSTR_FILEDESCRIPTORW);
 	clipboard->ID_FILECONTENTS = RegisterClipboardFormatW(CFSTR_FILECONTENTS);
 	clipboard->ID_PREFERREDDROPEFFECT = RegisterClipboardFormatW(CFSTR_PREFERREDDROPEFFECT);
 
-	clipboard->thread = CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE) cliprdr_thread_func, clipboard, 0, NULL);
+	if (!(clipboard->thread = CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE) cliprdr_thread_func, clipboard, 0, NULL)))
+		goto fail_clipboard_thread;
+
+	cliprdr->MonitorReady = wf_cliprdr_monitor_ready;
+	cliprdr->ServerCapabilities = wf_cliprdr_server_capabilities;
+	cliprdr->ServerFormatList = wf_cliprdr_server_format_list;
+	cliprdr->ServerFormatListResponse = wf_cliprdr_server_format_list_response;
+	cliprdr->ServerLockClipboardData = wf_cliprdr_server_lock_clipboard_data;
+	cliprdr->ServerUnlockClipboardData = wf_cliprdr_server_unlock_clipboard_data;
+	cliprdr->ServerFormatDataRequest = wf_cliprdr_server_format_data_request;
+	cliprdr->ServerFormatDataResponse = wf_cliprdr_server_format_data_response;
+	cliprdr->ServerFileContentsRequest = wf_cliprdr_server_file_contents_request;
+	cliprdr->ServerFileContentsResponse = wf_cliprdr_server_file_contents_response;
+	cliprdr->custom = (void*) wfc->clipboard;
+
+	return;
+
+fail_clipboard_thread:
+	CloseHandle(clipboard->req_fevent);
+	clipboard->req_fevent = NULL;
+fail_filecontents_event:
+	CloseHandle(clipboard->response_data_event);
+	clipboard->response_data_event = NULL;
+fail_data_event:
+	free(clipboard->fileDescriptor);
+	clipboard->fileDescriptor = NULL;
+fail_file_descriptor:
+	free(clipboard->file_names);
+	clipboard->file_names = NULL;
+fail_file_names:
+	free(clipboard->format_mappings);
+	clipboard->format_mappings = NULL;
+fail_format_mappings:
+	free(wfc->clipboard);
+	wfc->clipboard = NULL;
+
+	return;
```

Comment:
@hardening of course I thought about that but in that case it is completely useless if you look at the calling code ... unless rewriting the windows client which is not what this PR is about ;)